docs/proposals/container-init.md

```diff
@@ -0,0 +1,333 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Pod initialization
+
+@smarterclayton
+
+March 2016
+
+## Proposal and Motivation
+
+Within a pod there is a need to initialize local data or adapt to the current
+cluster environment that is not easily achieved in the current container model.
+Containers start in parallel after volumes are mounted, leaving no opportunity
+for coordination between containers without specialization of the image. If
+two containers need to share common initialization data, both images must
+be altered to cooperate using filesystem or network semantics, which introduces
+coupling between images. Likewise, if an image requires configuration in order
+to start and that configuration is environment dependent, the image must be
+altered to add the necessary templating or retrieval.
+
+This proposal introduces the concept of an **init container**, one or more
```

Comment:
If we want to do DAG, then we'd probably have data preconditions on
containers (the volume type container as proposed), and we'd need control
conditions.  I may have missed that in Brian's comment about doing volume
type container (which form the edges) as the implication of the change.

If we implement the DAG, init containers are technically still useful and
easier to reason about total resources for.  I believe in a future DAG
model you could convert init containers to a DAG defined solely in the
containers slice transparently, so one does not block the other.

On Thu, Mar 31, 2016 at 9:55 AM, Timothy St. Clair <notifications@github.com
> wrote:

> In docs/proposals/container-init.md
> <https://github.com/kubernetes/kubernetes/pull/23666#discussion_r58056775>
> :
>
> > +
> > +March 2016
> > +
> > +## Proposal and Motivation
> > +
> > +Within a pod there is a need to initialize local data or adapt to the current
> > +cluster environment that is not easily achieved in the current container model.
> > +Containers start in parallel after volumes are mounted, leaving no opportunity
> > +for coordination between containers without specialization of the image. If
> > +two containers need to share common initialization data, both images must
> > +be altered to cooperate using filesystem or network semantics, which introduces
> > +coupling between images. Likewise, if an image requires configuration in order
> > +to start and that configuration is environment dependent, the image must be
> > +altered to add the necessary templating or retrieval.
> > +
> > +This proposal introduces the concept of an **init container**, one or more
>
> This is a base case of DAG/workflow + pre&post-hook mechanics, except this
> proposal is missing the post bit.
>
> @bgrant0607 <https://github.com/bgrant0607>, @davidopp
> <https://github.com/davidopp> - Instead of speculating, I'll just ask ;-)
> Do you have a long term DAG/workflow model & hooks that you are planning?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23666/files/7efdc3d36197f1ca452339d49abbc87163fb0e75#r58056775>
>